Content.Server/Power/Generator/GasPowerProviderSystem.cs

```diff
@@ -0,0 +1,92 @@
+    [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+    [Dependency] private readonly NodeContainerSystem _nodeContainer = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<GasPowerProviderComponent, AtmosDeviceUpdateEvent>(OnDeviceUpdated);
+    }
+
+    private void OnDeviceUpdated(EntityUid uid, GasPowerProviderComponent component, AtmosDeviceUpdateEvent args)
+    {
+        if (component.LastProcess == TimeSpan.Zero)
+        {
+            component.LastProcess = _gameTiming.CurTime;
+            return; // Skip tick 0.
+        }
+
+        var timeDelta =  (float)(_gameTiming.CurTime - component.LastProcess).TotalSeconds;
+        component.LastProcess = _gameTiming.CurTime;
+
+        if (!HasComp<AtmosDeviceComponent>(uid)
+            || !TryComp<NodeContainerComponent>(uid, out var nodeContainer)
+            || !_nodeContainer.TryGetNode<PipeNode>(nodeContainer, "pipe", out var pipe))
+        {
+            return;
+        }
+
+        // if we're below the max temperature, then we are simply consuming our target gas
+        if (pipe.Air.Temperature <= component.MaxTemperature)
+        {
+            // we have enough gas, so we consume it and are powered
+            if (pipe.Air.Moles[(int) component.TargetGas] > component.MolesConsumedSec * timeDelta)
+            {
+                pipe.Air.AdjustMoles(component.TargetGas, -component.MolesConsumedSec * timeDelta);
+                SetPowered(uid, component, true);
+            }
+            else // we do not have enough gas, so we power off
+            {
+                SetPowered(uid, component, false);
+            }
+        }
+        else // we are exceeding the max temp and are now operating in pressure mode
+        {
+            var pres = component.PressureConsumedSec * timeDelta;
+            if (pipe.Air.Pressure >= pres)
+            {
+                // remove gas from the pipe
+                var res = pipe.Air.Remove(pres * 100.0f / (Atmospherics.R * pipe.Air.Temperature));
+                if (component.OffVentGas)
+                {
+                    // eject the gas into the atmosphere
+                    var mix = _atmosphereSystem.GetContainingMixture(uid, false, true);
+                    if (mix is not null)
+                        _atmosphereSystem.Merge(res, mix);
+                }
+
+                SetPowered(uid, component, true);
+            }
+            else // if we do not have high enough pressure to operate, power off
+            {
+                SetPowered(uid, component, false);
+            }
+        }
+    }
+
+    private void SetPowered(EntityUid uid, GasPowerProviderComponent comp, bool state)
+    {
+        if (state != comp.Powered)
+        {
+            comp.Powered = state;
+            var ev = new PowerChangedEvent(state, 0);
+            RaiseLocalEvent(uid, ref ev);
+        }
+    }
```

Comment:
Probably better off having an interface for IPowerReceivers and sharing it with ApcPowerReceiverComponent given they're similar in purpose.

Reply:
I'm not really sure how to do this in a sane way without diving into how some of the internals of power code.

Reply:
sloth is cring ignore them